File: MIDTERM-LAB-3/JS/MP07_SortRecords.js
/**
 * MP07 - Sort Records Alphabetically
 * Student: PARAON, JOSH HEIDRIC C.
 * Description: Prompts for a CSV file and sorts all records alphabetically 
 * by comparing every column in the dataset. Includes a retry loop for file input.
 */

const fs = require('fs');
const readline = require('readline/promises');

// Set up promise-based interface for async/await loop
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

/**
 * Parses a single CSV line, handling quoted fields.
 */
function parseCSVLine(line) {
    const tokens = [];
    let current = '';
    let inQuotes = false;

    for (const char of line) {
        if (char === '"') {
            inQuotes = !inQuotes;
        } else if (char === ',' && !inQuotes) {
            tokens.push(current);
            current = '';
        } else {
            current += char;
        }
    }
    tokens.push(current);
    return tokens;
}

async function main() {
    let lines = [];
    let headers = null;
    let fileLoaded = false;

    // Loop until a valid file is successfully loaded
    while (!fileLoaded) {
        const filePath = (await rl.question('Enter the CSV dataset file path: ')).trim();

        try {
            const fileContent = fs.readFileSync(filePath, 'utf8');
            lines = fileContent.split('\n').filter(line => line.trim() !== '');

            if (lines.length === 0) {
                console.log('Error: The file is empty. Please try a different file.');
                continue;
            }

            headers = parseCSVLine(lines[0]);
            fileLoaded = true; // Break the loop

        } catch (err) {
            if (err.code === 'ENOENT') {
                console.log(`Error: File not found at '${filePath}'. Please try again.`);
            } else {
                console.log(`Error reading file: ${err.message}`);
            }
        }
    }

    // Step 5: Parse data rows
    const records = lines.slice(1).map(line => parseCSVLine(line));

    // Sort records alphabetically by comparing the entire row content
    records.sort((row1, row2) => {
        const minLength = Math.min(row1.length, row2.length);
        
        // Compare column by column
        for (let i = 0; i < minLength; i++) {
            const val1 = row1[i].trim();
            const val2 = row2[i].trim();
            const comparison = val1.localeCompare(val2, undefined, { sensitivity: 'base' });

            // If columns are different, we have our alphabetical order
            if (comparison !== 0) {
                return comparison;
            }
        }
        // Tie-breaker: shorter row comes first if all common columns match
        return row1.length - row2.length;
    });

    // Step 6: Display formatted output
    console.log('\n========================================');
    console.log('Records Sorted: Alphabetically by Entire Dataset');
    console.log(`Total Records: ${records.length}`);
    console.log('========================================');

    // Print header
    console.log(headers.map(h => h.trim()).join(' | '));
    console.log('-'.repeat(60));

    // Print each alphabetically sorted record
    records.forEach((row, index) => {
        console.log(`[${index + 1}] ${row.map(f => f.trim()).join(' | ')}`);
    });

    rl.close();
}

main();

File: MIDTERM-LAB-3/Java/MP07_SortRecords.java
import java.io.*;
import java.util.*;

/**
 * MP07 - Sort Entire Dataset Alphabetically
 * Student: PARAON, JOSH HEIDRIC C.
 * Description: Prompts for a CSV file and sorts all records alphabetically 
 * by comparing every column in the dataset. Includes a retry loop for file input.
 */
public class MP07_SortRecords {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<String[]> records = new ArrayList<>();
        String[] headers = null;
        boolean fileLoaded = false;

        // Loop until a valid file is provided
        while (!fileLoaded) {
            System.out.print("Enter the CSV dataset file path: ");
            String filePath = scanner.nextLine().trim();

            try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
                String line;
                boolean isHeader = true;

                // Clear previous attempts if any
                records.clear();

                while ((line = reader.readLine()) != null) {
                    String[] fields = parseCSVLine(line);
                    if (isHeader) {
                        headers = fields;
                        isHeader = false;
                    } else {
                        records.add(fields);
                    }
                }

                if (headers == null) {
                    System.out.println("Error: The file is empty. Please try a different file.");
                } else {
                    fileLoaded = true; // Success! Break the loop
                }

            } catch (FileNotFoundException e) {
                System.out.println("Error: File not found at '" + filePath + "'. Please check the path and try again.");
            } catch (IOException e) {
                System.out.println("Error reading file: " + e.getMessage());
            }
        }

        // Sort records alphabetically by comparing the entire row content
        records.sort((row1, row2) -> {
            int minLength = Math.min(row1.length, row2.length);
            for (int i = 0; i < minLength; i++) {
                int comparison = row1[i].trim().compareToIgnoreCase(row2[i].trim());
                if (comparison != 0) return comparison;
            }
            return Integer.compare(row1.length, row2.length);
        });

        // Display the sorted records
        System.out.println("\n========================================");
        System.out.println("Records Sorted: Alphabetically by Entire Dataset");
        System.out.println("Total Records: " + records.size());
        System.out.println("========================================");

        // Print header row
        System.out.println(String.join(" | ", headers));
        System.out.println("-".repeat(60));

        // Print each alphabetically sorted record
        for (int i = 0; i < records.size(); i++) {
            String[] row = records.get(i);
            System.out.printf("[%d] %s%n", i + 1, String.join(" | ", row));
        }

        scanner.close();
    }

    /**
     * Parses a single CSV line handling quoted fields.
     */
    static String[] parseCSVLine(String line) {
        List<String> tokens = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;

        for (char c : line.toCharArray()) {
            if (c == '"') {
                inQuotes = !inQuotes;
            } else if (c == ',' && !inQuotes) {
                tokens.add(current.toString());
                current.setLength(0);
            } else {
                current.append(c);
            }
        }
        tokens.add(current.toString());
        return tokens.toArray(new String[0]);
    }
}